Source/Core/Core/PowerPC/CachedInterpreter/CachedInterpreter.h

```diff
@@ -30,32 +43,108 @@ class CachedInterpreter : public JitBase
   const char* GetName() const override { return "Cached Interpreter"; }
   const CommonAsmRoutinesBase* GetAsmRoutines() override { return nullptr; }
 
 private:
-  struct Instruction;
-
-  u8* GetCodePtr();
   void ExecuteOneBlock();
 
   bool HandleFunctionHooking(u32 address);
 
-  static void EndBlock(CachedInterpreter& cached_interpreter, UGeckoInstruction data);
-  static void UpdateNumLoadStoreInstructions(CachedInterpreter& cached_interpreter,
-                                             UGeckoInstruction data);
-  static void UpdateNumFloatingPointInstructions(CachedInterpreter& cached_interpreter,
-                                                 UGeckoInstruction data);
-  static void WritePC(CachedInterpreter& cached_interpreter, UGeckoInstruction data);
-  static void WriteBrokenBlockNPC(CachedInterpreter& cached_interpreter, UGeckoInstruction data);
-  static bool CheckFPU(CachedInterpreter& cached_interpreter, u32 data);
-  static bool CheckDSI(CachedInterpreter& cached_interpreter, u32 data);
-  static bool CheckProgramException(CachedInterpreter& cached_interpreter, u32 data);
-  static bool CheckBreakpoint(CachedInterpreter& cached_interpreter, u32 data);
-  static bool CheckIdle(CachedInterpreter& cached_interpreter, u32 idle_pc);
-
-  BlockCache m_block_cache{*this};
-  std::vector<Instruction> m_code;
+  // Finds a free memory region and sets the code emitter to point at that region.
+  // Returns false if no free memory region can be found.
+  bool SetEmitterStateToFreeCodeRegion();
+
+  void FreeRanges();
+  void ResetFreeMemoryRanges();
+
+  struct EndBlockOperands;
+  template <bool check_exceptions>
+  struct InterpretOperands;
+  struct HLEFunctionOperands;
+  struct WriteBrokenBlockNPCOperands;
+  struct CheckFPUOperands;
+  struct CheckBreakpointOperands;
+  struct CheckIdleOperands;
+
+  static s32 EndBlock(PowerPC::PowerPCState& ppc_state, const EndBlockOperands& operands);
+  template <bool check_exceptions, bool write_pc>
+  static s32 Interpret(PowerPC::PowerPCState& ppc_state,
+                       const InterpretOperands<check_exceptions>& operands);
+  static s32 HLEFunction(PowerPC::PowerPCState& ppc_state, const HLEFunctionOperands& operands);
+  static s32 WriteBrokenBlockNPC(PowerPC::PowerPCState& ppc_state,
+                                 const WriteBrokenBlockNPCOperands& operands);
+  static s32 CheckFPU(PowerPC::PowerPCState& ppc_state, const CheckFPUOperands& operands);
+  static s32 CheckBreakpoint(PowerPC::PowerPCState& ppc_state,
+                             const CheckBreakpointOperands& operands);
+  static s32 CheckIdle(PowerPC::PowerPCState& ppc_state, const CheckIdleOperands& operands);
+
+  HyoutaUtilities::RangeSizeSet<u8*> m_free_ranges;
+  CachedInterpreterBlockCache m_block_cache;
+};
+
+struct CachedInterpreter::EndBlockOperands
+{
+  u32 downcount;
+  u32 num_load_stores;
+  u32 num_fp_inst;
+  u32 : 32;
+};
+
+template <>
+struct CachedInterpreter::InterpretOperands<false>
```

Comment:
If you need a template argument for this struct, I'd say go ahead and use two separate structs instead. You'll have to have two copies of the `Interpret` source code, but the amount of code in `Interpret` is pretty small when not counting the exception handling code, so I think the increase in clarity would be worth it.

Reply:
Not feasible. I think what you are envisioning would look like this:
```cpp
template <bool write_pc>
s32 CachedInterpreter::Interpret<false, write_pc>(PowerPC::PowerPCState& ppc_state,
                                                  const InterpretOperandsFalse& operands)
{
  if constexpr (write_pc)
  {
    ppc_state.npc = (ppc_state.pc = operands.current_pc) + 4;
  }
  operands.func(operands.interpreter, operands.inst);
  return sizeof(AnyCallback) + sizeof(operands);
}

template <bool write_pc>
s32 CachedInterpreter::Interpret<true, write_pc>(PowerPC::PowerPCState& ppc_state,
                                                 const InterpretOperandsTrue& operands)
{
  if constexpr (write_pc)
  {
    ppc_state.npc = (ppc_state.pc = operands.current_pc) + 4;
  }
  operands.func(operands.interpreter, operands.inst);
  if ((ppc_state.Exceptions & (EXCEPTION_DSI | EXCEPTION_PROGRAM)) != 0)
  {
    ppc_state.pc = operands.current_pc;
    ppc_state.downcount -= operands.downcount;
    operands.power_pc.CheckExceptions();
    return 0;
  }
  return sizeof(AnyCallback) + sizeof(operands);
}
```
Function template partial specialization is not allowed in C++, so in order to turn `InterpretOperands` into two separate structs I would need to create *four* separate definitions for `Interpret`.

One potential improvement for how it is now would be to make `CachedInterpreter::InterpretOperands<true>` derive from its `false` counterpart to be more DRY.

Reply:
You could have two functions with different names, with each having `write_pc` as its only template parameter.

Reply:
I definitely preferred it the way I had it, but I've changed it to your suggestion.